### PR TITLE
Consider VM CPU and Memory requirements for POD scheduling

### PR DESCRIPTION
**Pull request by vladikr**
The following changes will allow users to express CPU and Memory requirements and constraints for scheduling purposes.  Such as:

```yaml
    resources:
   requests:
        memory: "1Gi"
        cpu: "1"
      limits:
        memory: "2Gi"
        cpu: "2"
```
 
The VM requirements will be translated to a POD resources request, including a VM memory overhead that will be calculated and added to the POD memory request.

### CODE REVIEW

File: pkg/api/v1/schema.go [hunk 1]
```diff
@@ -66,6 +66,10 @@ type ResourceRequirements struct {
 	// Valid resource keys are "memory" and "cpu".
 	// +optional
 	Requests v1.ResourceList `json:"requests,omitempty"`
+	// Limits describes the maximum amount of compute resources allowed.
+	// Valid resource keys are "memory" and "cpu".
```
Comment (stu-gott) [inline pkg/api/v1/schema.go]: You mention memory as a valid resource, but the rest of the PR seems to specifically address CPU, did I miss something? :)

Comment (vladikr) [reply pkg/api/v1/schema.go]: > You mention memory as a valid resource, but the rest of the PR seems to specifically address CPU, did I miss something? :)
You're right, I'll add it in the next PR that handles the memory limits. :)

**Comment by davidvossel** [comment]
@fabiand 

It isn't obvious to me why we're exposing the "limits" value to the user on the VirtualMachine spec when it's only a value that's being applied to the VM Pod and not the libvirt domain.

If "limits" are only being used on the pod, could we not remove that value from the user and calculate limits as VM requests + overhead?

To rephrase this in another way, what value are we providing to the user by exposing the "limits" option? I can't answer that clearly right now.

**Comment by rmohr** [comment]
> It isn't obvious to me why we're exposing the "limits" value to the user on the VirtualMachine spec when it's only a value that's being applied to the VM Pod and not the libvirt domain.

Shouldn't that assign a cpu share to the cpu-cgroup where the vm runs in, since the vm runs in that pod? Sounds useful to me.

> If "limits" are only being used on the pod, could we not remove that value from the user and calculate limits as VM requests + overhead?

I think that would give our "requests" on the vm a different meaning than "requests" on the pod then.

File: pkg/virt-controller/services/template.go [hunk 2]
```diff
@@ -73,6 +73,39 @@ func (t *templateService) RenderLaunchManifest(vm *v1.VirtualMachine) (*kubev1.P
 	// the qemu pid is in the pods mount namespace
 	privateDir := fmt.Sprintf("%s-private/%s/%s", t.virtShareDir, namespace, domain)
 
+	// Consider CPU and memory requests and limits for pod scheduling
+	resources := kubev1.ResourceRequirements{}
+	vmResources := vm.Spec.Domain.Resources
+	requests := kubev1.ResourceList{}
```
Comment (rmohr) [inline pkg/virt-controller/services/template.go]: I think that you can just directly attach requests and limits to the resources. No need for the temporary maps. It should not map for the serializer if they are empty or nil.

**Comment by davidvossel** [comment]
> Shouldn't that assign a cpu share to the cpu-cgroup where the vm runs in, since the vm runs in that pod? Sounds useful to me.

Okay, so for example, a user may want to give a VM 1 core, but limit the cpu share to only consume 1/2 a core's cycles.  They'd do this by setting the cpu limit?

**Comment by rmohr** [comment]
> Okay, so for example, a user may want to give a VM 1 core, but limit the cpu share to only consume 1/2 a core's cycles. They'd do this by setting the cpu limit?

Yes, one can say, my single guest core should e.g. get the resource maximum representing 0.5 cpu cores on the node. Requests would be the most important value for the scheduler, limits is really a limit on the node. We may or may not check in the future, that the limits and the guest core number make sense. E.g. having one core (one process) and assigning it 3.2 host cores does not really improve the performance of the one guest core.

File: pkg/virt-controller/services/template.go [hunk 3]
```diff
@@ -194,6 +218,52 @@ func (t *templateService) RenderLaunchManifest(vm *v1.VirtualMachine) (*kubev1.P
...
+	overhead.Add(*pagetableMemory)
+
+	// Add fixed overhead for shared libraries and such
+	// TODO account for the overhead of kubevirt components running in the pod
+	overhead.Add(*resource.NewScaledQuantity(64, resource.Mega))
+
+	// Add CPU table overhead (8 MiB per vCPU and 8 MiB per IO thread)
+	// TODO
+
+	/* Add video RAM overhead
```
Comment (vladikr) [inline pkg/virt-controller/services/template.go]: I wonder how users can now request a specific video device to be attached? I see that 7e17db9c257410796bfdf06b0084cf9189b7c933 removed it from the vm spec.

Comment (rmohr) [reply pkg/virt-controller/services/template.go]: Right now we always attach one.

File: pkg/virt-controller/services/template.go [hunk 4]
```diff
@@ -194,6 +218,52 @@ func (t *templateService) RenderLaunchManifest(vm *v1.VirtualMachine) (*kubev1.P
...
+	overhead := resource.NewScaledQuantity(0, resource.Kilo)
+
+	// Add the memory needed for pagetables (one bit for every 512b of RAM size)
+	pagetableMemory := resource.NewScaledQuantity(vmMemoryReq.ScaledValue(resource.Kilo), resource.Kilo)
+	pagetableMemory.Set(pagetableMemory.Value() / 512)
+	overhead.Add(*pagetableMemory)
+
+	// Add fixed overhead for shared libraries and such
+	// TODO account for the overhead of kubevirt components running in the pod
+	overhead.Add(*resource.NewScaledQuantity(64, resource.Mega))
```
Comment (rmohr) [inline pkg/virt-controller/services/template.go]: I find `resource.MustParse("64M")` (64MB) easy to read.

File: pkg/virt-controller/services/template.go (hunk 3 above)
Comment (rmohr) [reply pkg/virt-controller/services/template.go]: It is a static overhead of `resource.MustParse("16Mi")` right now.

File: pkg/virt-controller/services/template.go [hunk 5]
```diff
@@ -194,6 +218,52 @@ func (t *templateService) RenderLaunchManifest(vm *v1.VirtualMachine) (*kubev1.P
...
+
+	// Add CPU table overhead (8 MiB per vCPU and 8 MiB per IO thread)
```
Comment (rmohr) [inline pkg/virt-controller/services/template.go]: Right now we only support setting vCPUs. IO threads and sockets is one. @MarSik does that mean that we have to do `8Mi + 8Mi` for one cpu with one trhead?

File: pkg/virt-controller/services/template.go [hunk 6]
```diff
@@ -210,6 +234,50 @@ func (t *templateService) RenderLaunchManifest(vm *v1.VirtualMachine) (*kubev1.P
...
+	pagetableMemory.Set(pagetableMemory.Value() / 512)
+	overhead.Add(*pagetableMemory)
+
+	// Add fixed overhead for shared libraries and such
+	// TODO account for the overhead of kubevirt components running in the pod
+	overhead.Add(resource.MustParse("64M"))
+
+	// Add CPU table overhead (8 MiB per vCPU and 8 MiB per IO thread)
+	// Currently, will add a static overhead for vCPU+IOThread
+	overhead.Add(resource.MustParse("16Mi"))
```
Comment (rmohr) [inline pkg/virt-controller/services/template.go]: I think that you need to multiply that with the amount of cores.

**Comment by vladikr** [comment]
> Also an integration test is missing. Maybe adding a "By()" to the multi-core vm test in configurations_test.go is a good enough start (having a test for one node and one for more than one would be great).

I think integration test will be tricky...
Inside the guest, we will never get the same amount of memory we've configured, because of kernel reservations. Looking at the qemu process won't help us much because libvirt does its own adjustments (I think due to q35), so if we'll configure 64MB the qemu process will be setup with -m62

**Comment by rmohr** [comment]
@vladikr I just thought about checking the created and running launcher pod.

File: pkg/virt-controller/services/template.go [hunk 7]
```diff
@@ -209,6 +233,57 @@ func (t *templateService) RenderLaunchManifest(vm *v1.VirtualMachine) (*kubev1.P
...
+	overhead.Add(*pagetableMemory)
+
+	// Add fixed overhead for shared libraries and such
+	// TODO account for the overhead of kubevirt components running in the pod
+	overhead.Add(resource.MustParse("64M"))
+
+	// Add CPU table overhead (8 MiB per vCPU and 8 MiB per IO thread)
+	// overhead per vcpu in MiB
+	coresMemory := uint32(8)
+	if domain.CPU != nil {
```
Comment (rmohr) [inline pkg/virt-controller/services/template.go]: I think I missed setting the default values on the CPU if it was not specified in the defaulter.

**Comment by cynepco3hahue** [comment]
@vladikr We can check kernel memory usage by running
`dmesg |grep Memory:` and sum it with `cat /proc/meminfo | grep MemTotal:` to get total VM memory.

**Comment by vladikr** [comment]
> @vladikr We can check kernel memory usage by running
> dmesg |grep Memory: and sum it with cat /proc/meminfo | grep MemTotal: to get total VM memory.

You're right, but that will never give us the same number that was defined in the libvirt xml for memory, due to kernel reservations. This makes it harder to test it in a functional test.

**Comment by rmohr** [comment]
Testing if we have the right value on the pod should be good enough for the start.

**Pull request merged**